File: src/main/java/QueryHandler.java
import java.util.List;
import java.util.HashSet;

public class QueryHandler {
    private InvertedIndex index;

    public QueryHandler(InvertedIndex index){
        this.index = index;
    }

    public void setIndex(InvertedIndex index) {
        this.index = index;
    }

    public HashSet<String> handleQuery(Query query){
        HashSet<String> universalSet = new HashSet<>(index.getDocIdToContents().keySet());
        HashSet<String> answer = getIntersectionSet(query.getAndWords(), universalSet);
        if(!query.getOrWords().isEmpty()){
            answer.retainAll(getUnionSet(query.getOrWords()));
        }
        answer.removeAll(getUnionSet(query.getNotWords()));
        return answer;
    }

    private HashSet<String> getUnionSet(List<String> words){
        HashSet<String> set = new HashSet<>();
        for(String word : words){
            set.addAll(index.search(word));
        }
        return set;
    }

    private HashSet<String> getIntersectionSet(List<String> words, HashSet<String> universalSet){
        HashSet<String> set = new HashSet<>(universalSet);
        for(String word : words){
            set.retainAll(index.search(word));
        }
        return set;
    }
}

File: Phase05/InvertedIndex/QueryHandler.cs
using InvertedIndex.Abstraction;

namespace InvertedIndex;

public class QueryHandler : IHandler
{

    private readonly ISearchEngine _engine;

    public QueryHandler(ISearchEngine engine)
    {
        _engine = engine;
    }

    public HashSet<string> HandleQuery(Query query)
    {
        var universalSet = new HashSet<string>(_engine.DocIdToContents.Keys);
        var answer = GetIntersectionSet(query.AndWords, universalSet);
        if (query.OrWords.Any())
        {
            answer.IntersectWith(GetUnionSet(query.OrWords));
        }

        answer.RemoveWhere(p => GetUnionSet(query.NotWords).Contains(p));
        return answer;
    }

    private HashSet<string> GetUnionSet(List<string> words)
    {
        var set = new HashSet<string>();
        foreach (var word in words)
        {
            set.UnionWith(_engine.SearchForWord(word));
        }

        return set;
    }

    private HashSet<string> GetIntersectionSet(List<string> words, HashSet<string> universalSet)
    {
        var set = new HashSet<string>(universalSet);
        foreach (string word in words)
        {
            set.IntersectWith(_engine.SearchForWord(word));
        }
        return set;
    }
}